{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re \n",
    "import itertools\n",
    "\n",
    "def idx_to_ranges(i):\n",
    "    for a, b in itertools.groupby(enumerate(i), lambda pair: pair[1] - pair[0]):\n",
    "        b = list(b)\n",
    "        yield b[0][1], b[-1][1]\n",
    "\n",
    "def get_blank_regions(seq, _blankchar='~'):\n",
    "    _indices = [m.start() for m in re.finditer(_blankchar,seq)]\n",
    "    return list(idx_to_ranges(_indices))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Import aligned sequences and input features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_header</th>\n",
       "      <th>aligned_sequence</th>\n",
       "      <th>pt_code</th>\n",
       "      <th>isolate_name</th>\n",
       "      <th>sample_date</th>\n",
       "      <th>genes</th>\n",
       "      <th>country</th>\n",
       "      <th>source</th>\n",
       "      <th>arv_status</th>\n",
       "      <th>_merge</th>\n",
       "      <th>gap_windows</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;PT1000 IS1001 Integrase naive</td>\n",
       "      <td>tttttagatgggatagataaagcccaagaagaacatgaaaaatacc...</td>\n",
       "      <td>PT1000</td>\n",
       "      <td>IS1001</td>\n",
       "      <td>4/7/2017</td>\n",
       "      <td>PR/RT</td>\n",
       "      <td>Italy</td>\n",
       "      <td>plasma</td>\n",
       "      <td>naive</td>\n",
       "      <td>both</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;PT2000 IS2001 Integrase naive</td>\n",
       "      <td>tttttagatggaatagataaggcacaagaagarcatgagaaatatc...</td>\n",
       "      <td>PT1000</td>\n",
       "      <td>IS1001</td>\n",
       "      <td>4/7/2017</td>\n",
       "      <td>Integrase</td>\n",
       "      <td>Italy</td>\n",
       "      <td>plasma</td>\n",
       "      <td>naive</td>\n",
       "      <td>both</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;PT3000 IS3001 Integrase experienced</td>\n",
       "      <td>tttttagatggaatagayaaggcccaagaagaacatgagaagtatc...</td>\n",
       "      <td>PT2000</td>\n",
       "      <td>IS2001</td>\n",
       "      <td>4/6/2016</td>\n",
       "      <td>PR/RT</td>\n",
       "      <td>Italy</td>\n",
       "      <td>plasma</td>\n",
       "      <td>naive</td>\n",
       "      <td>both</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;PT4000 IS4001 Integrase experienced</td>\n",
       "      <td>tttctagatggaatagataaggcccaagaagatcatgagaaatatc...</td>\n",
       "      <td>PT2000</td>\n",
       "      <td>IS2001</td>\n",
       "      <td>4/6/2016</td>\n",
       "      <td>Integrase</td>\n",
       "      <td>Italy</td>\n",
       "      <td>plasma</td>\n",
       "      <td>naive</td>\n",
       "      <td>both</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;PT1000 IS1001 PR/RT naive</td>\n",
       "      <td>cctcagatcactctttggcaacgacccatcgtcacagtaagggtag...</td>\n",
       "      <td>PT3000</td>\n",
       "      <td>IS3001</td>\n",
       "      <td>12/12/2012</td>\n",
       "      <td>PR/RT</td>\n",
       "      <td>Italy</td>\n",
       "      <td>plasma</td>\n",
       "      <td>Experienced</td>\n",
       "      <td>both</td>\n",
       "      <td>[(297, 302)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>&gt;PT2000 IS2001 PR/RT naive</td>\n",
       "      <td>cctcagatcactctttggcaacgacccctcgtcacaataaagatag...</td>\n",
       "      <td>PT3000</td>\n",
       "      <td>IS3001</td>\n",
       "      <td>12/12/2012</td>\n",
       "      <td>Integrase</td>\n",
       "      <td>Italy</td>\n",
       "      <td>plasma</td>\n",
       "      <td>Experienced</td>\n",
       "      <td>both</td>\n",
       "      <td>[(297, 302)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>&gt;PT3000 IS3001 PR/RT experienced</td>\n",
       "      <td>cctcaratcactctttggcaacgacccrtygtcacagtaagggtag...</td>\n",
       "      <td>PT4000</td>\n",
       "      <td>IS4001</td>\n",
       "      <td>9/3/2009</td>\n",
       "      <td>PR/RT</td>\n",
       "      <td>Italy</td>\n",
       "      <td>plasma</td>\n",
       "      <td>Experienced</td>\n",
       "      <td>both</td>\n",
       "      <td>[(297, 302)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>&gt;PT4000 IS4001 PR/RT experienced</td>\n",
       "      <td>cctcaaatcactctttggcaacgacctattgtcacagtaaagatag...</td>\n",
       "      <td>PT4000</td>\n",
       "      <td>IS4001</td>\n",
       "      <td>9/3/2009</td>\n",
       "      <td>Integrase</td>\n",
       "      <td>Italy</td>\n",
       "      <td>plasma</td>\n",
       "      <td>Experienced</td>\n",
       "      <td>both</td>\n",
       "      <td>[(297, 302)]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        sequence_header  \\\n",
       "0        >PT1000 IS1001 Integrase naive   \n",
       "1        >PT2000 IS2001 Integrase naive   \n",
       "2  >PT3000 IS3001 Integrase experienced   \n",
       "3  >PT4000 IS4001 Integrase experienced   \n",
       "4            >PT1000 IS1001 PR/RT naive   \n",
       "5            >PT2000 IS2001 PR/RT naive   \n",
       "6      >PT3000 IS3001 PR/RT experienced   \n",
       "7      >PT4000 IS4001 PR/RT experienced   \n",
       "\n",
       "                                    aligned_sequence pt_code isolate_name  \\\n",
       "0  tttttagatgggatagataaagcccaagaagaacatgaaaaatacc...  PT1000       IS1001   \n",
       "1  tttttagatggaatagataaggcacaagaagarcatgagaaatatc...  PT1000       IS1001   \n",
       "2  tttttagatggaatagayaaggcccaagaagaacatgagaagtatc...  PT2000       IS2001   \n",
       "3  tttctagatggaatagataaggcccaagaagatcatgagaaatatc...  PT2000       IS2001   \n",
       "4  cctcagatcactctttggcaacgacccatcgtcacagtaagggtag...  PT3000       IS3001   \n",
       "5  cctcagatcactctttggcaacgacccctcgtcacaataaagatag...  PT3000       IS3001   \n",
       "6  cctcaratcactctttggcaacgacccrtygtcacagtaagggtag...  PT4000       IS4001   \n",
       "7  cctcaaatcactctttggcaacgacctattgtcacagtaaagatag...  PT4000       IS4001   \n",
       "\n",
       "  sample_date      genes country  source   arv_status _merge   gap_windows  \n",
       "0    4/7/2017      PR/RT   Italy  plasma        naive   both            []  \n",
       "1    4/7/2017  Integrase   Italy  plasma        naive   both            []  \n",
       "2    4/6/2016      PR/RT   Italy  plasma        naive   both            []  \n",
       "3    4/6/2016  Integrase   Italy  plasma        naive   both            []  \n",
       "4  12/12/2012      PR/RT   Italy  plasma  Experienced   both  [(297, 302)]  \n",
       "5  12/12/2012  Integrase   Italy  plasma  Experienced   both  [(297, 302)]  \n",
       "6    9/3/2009      PR/RT   Italy  plasma  Experienced   both  [(297, 302)]  \n",
       "7    9/3/2009  Integrase   Italy  plasma  Experienced   both  [(297, 302)]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Import from spreadsheet ###\n",
    "aligned_sequences = pd.read_csv('aligned_sequences.csv')\n",
    "input_features = pd.read_csv('input_features.csv')\n",
    "# display(aligned_sequences, input_features)\n",
    "\n",
    "##### Merge back in addtl info #####\n",
    "merged_df = pd.merge(aligned_sequences,input_features,on='sequence_header',how='outer',indicator=True)\n",
    "# display(aligned, aligned['_merge'].value_counts())\n",
    "\n",
    "##### Get gap windows #####\n",
    "# lower aligned_seq #\n",
    "merged_df['aligned_sequence'] = merged_df['aligned_sequence'].str.lower()\n",
    "# Left strip blanks #\n",
    "merged_df['aligned_sequence'] = merged_df['aligned_sequence'].str.lstrip('-')\n",
    "# Right strip blanks #\n",
    "merged_df['aligned_sequence'] = merged_df['aligned_sequence'].str.rstrip('-')\n",
    "# replace '-' -> '~' #\n",
    "merged_df['aligned_sequence'] = merged_df['aligned_sequence'].str.replace('-','~')\n",
    "# get gap windows #\n",
    "merged_df['gap_windows'] = merged_df['aligned_sequence'].apply(lambda x: get_blank_regions(x))\n",
    "# replace ~ w/ 'n' #\n",
    "merged_df['aligned_sequence'] = merged_df['aligned_sequence'].str.replace('~','n')\n",
    "\n",
    "##### Keep relevant columns #####\n",
    "display(merged_df)\n",
    "\n",
    "### Inspection ###\n",
    "# merged_df.to_csv('merged_df.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Create FASTA file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Export #####\n",
    "with open(f'bankit_upload.fasta', 'w') as f:\n",
    "    for index, row in merged_df.iterrows():\n",
    "        _sample_date = row['sample_date']\n",
    "        _country = row['country']\n",
    "        _pt_code = row['pt_code']\n",
    "        _isolate_name = row['isolate_name']\n",
    "        _genes = row['genes']\n",
    "        _arv_status = row['arv_status']\n",
    "        _bankit_isolate=f'{_pt_code}_{_isolate_name}_{_genes}_{_arv_status}'\n",
    "        _aligned_seq = row['aligned_sequence']\n",
    "        \n",
    "        f.write(f'>{_bankit_isolate}')\n",
    "        f.write(f'[Collection_date={_sample_date}]')\n",
    "        f.write(f'[Country={_country}]')\n",
    "        f.write(f'[Host=Homo sapiens]')\n",
    "        f.write(f'[Organism=Human immunodeficiency virus 1]')\n",
    "        f.write(f'[Isolate={_bankit_isolate}]')\n",
    "        f.write(f'[Isolation source=Plasma]')\n",
    "        f.write(f'\\n')\n",
    "        f.write(f'{_aligned_seq}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Create features file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "_stop_codon_seqs = [\n",
    "    '>PT1000_IS1001_Integrase_naive',\n",
    "    '>PT2000_IS2001_PR/RT_naive'\n",
    "]\n",
    "\n",
    "_misc_note_prrt = 'nonfunctional pol protein due to mutation; contains protease and reverse transcriptase'\n",
    "_misc_note_in = 'nonfunctional pol protein due to mutation; contains integrase'\n",
    "\n",
    "\n",
    "with open('features.txt', 'w') as f:\n",
    "    for index, row in merged_df.iterrows():\n",
    "        # Feature, sequence#\n",
    "        _pt_code = row['pt_code']\n",
    "        _isolate_name = row['isolate_name']\n",
    "        _genes = row['genes']\n",
    "        _arv_status = row['arv_status']\n",
    "        _bankit_isolate=f'{_pt_code}_{_isolate_name}_{_genes}_{_arv_status}'\n",
    "        f.write(f'>Feature {_bankit_isolate}\\n')\n",
    "        _end_position = len(row['aligned_sequence'])\n",
    "\n",
    "        # gene #\n",
    "        f.write(f'<1\\t>{_end_position}\\tgene\\n')\n",
    "        f.write(f'\\t\\t\\tgene\\tpol\\n')\n",
    "\n",
    "        for gap in row['gap_windows']:\n",
    "            _gap_start = gap[0]\n",
    "            _gap_end = gap[1]\n",
    "            f.write(f'{_gap_start}\\t{_gap_end}\\tgap\\n')\n",
    "            f.write(f'\\t\\t\\testimated_length\\t{_gap_end-_gap_start+1}\\n')\n",
    "\n",
    "        if _isolate_name in _stop_codon_seqs:\n",
    "            f.write(f'<1\\t>{_end_position}\\tmisc_feature\\n')\n",
    "            if row['genes'] == 'PR/RT':\n",
    "                f.write(f'\\t\\t\\tnote\\t{_misc_note_prrt}\\n')\n",
    "            if row['genes'] == 'Integrase':\n",
    "                f.write(f'\\t\\t\\tnote\\t{_misc_note_in}\\n')\n",
    "\n",
    "        # CDS #\n",
    "        f.write(f'<1\\t>{_end_position}\\tCDS\\n')\n",
    "        f.write(f'\\t\\t\\tgene\\tpol\\n')\n",
    "        f.write(f'\\t\\t\\tcodon_start\\t1\\n')\n",
    "        f.write(f'\\t\\t\\tproduct\\tpol protein\\n')\n",
    "        f.write(f'\\t\\t\\ttransl_table\\t1\\n')\n",
    "        # end #\n",
    "        f.write(f'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.12 ('hivdb')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "868b1c0de0534bb9441798eecc360b31a36a015181c76fcaa0d6d664337e4318"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
